Feature/10058 check if product already blazed

JorgeMucientes opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #10058 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->
Hides Blaze product CTA on product detail screen if the current product has an active ongoing campaign. Note that for this use case we consider a campaign to be active if it's in one of the following states:
- InModeration
- Scheduled
- Active

### Testing instructions
<!-- Step by step testing instructions. When necessary break out individual scenarios that need testing, consider including a checklist for the reviewer to go through. -->
1. Log into a WP.com site with at last one product published
2. Navigate to product detail and click on "Promote with Blaze" 
3. Create a campaign
4. Navigate again to the same product detail and notice how the CTA is gone

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->

https://github.com/woocommerce/woocommerce-android/assets/2663464/d18caad2-34e2-4bb4-98b7-cdcf44c3985b

JorgeMucientes committed:
Check if a product is already being promoted with Blaze

If the product is already promoted we hide the Blaze cta inside product detail

JorgeMucientes committed:
Fix unit tests compile issues

JorgeMucientes committed:
Suppress complex condition as the condition is readable enough

JorgeMucientes committed:
Update FluxC changeset

0nko commented:
I've tried testing your PR but I'm seeing a weird empty state when initiating blaze. Do I need to enable something on the site? I'm getting the same behaviour on multiple sites.

[blaze_empty.webm](https://github.com/woocommerce/woocommerce-android/assets/1522856/55945ecc-f7cb-4b10-a217-7f51daf976ec)

I see this in the logs:
```
2023-11-02 19:17:29.297  1032-1032  chromium                com.woocommerce.android.dev          I  [INFO:CONSOLE(28)] "Uncaught TypeError: window.AppBoot is not a function", source: https://wordpress.com/advertising/woowoozela.com?blazepress-widget=post-7529&_source=my_store_section (28)
2023-11-02 19:17:29.297  1032-1032  chromium                com.woocommerce.android.dev          I  [INFO:CONSOLE(31)] "%cSTOP! color:#f00;font-size:xx-large", source: https://wordpress.com/advertising/woowoozela.com?blazepress-widget=post-7529&_source=my_store_section (31)
2023-11-02 19:17:29.297  1032-1032  chromium                com.woocommerce.android.dev          I  [INFO:CONSOLE(32)] "%cWait! This browser feature runs code that can alter your website or its security, and is intended for developers. If you've been told to copy and paste something here to enable a feature, someone may be trying to compromise your account. Please make sure you understand the code and trust the source before adding anything here. font-size:large;", source: https://wordpress.com/advertising/woowoozela.com?blazepress-widget=post-7529&_source=my_store_section (32)
```

JorgeMucientes committed:
Make targetUrn nullable

Target urn is not a critical field to work with campaign data and might be null when upgrading the app from older versions that didn't save this value into de database

JorgeMucientes commented:
Thanks for reviewing this @0nko. 

> I'm seeing a weird empty state when initiating blaze. Do I need to enable something on the site? I'm getting the same behaviour on multiple sites.

That is very weird and afaik completely unrelated to the changes. I've experienced that bug from `trunk` too, after doing a clean install of the app (clearing app's storage). I will investigate more but I think is completely unrelated to these changes. 

However, I did notice a crash on the database migration strategy when transitioning from an installed version of the app from `trunk` with persisted campaigns and then installing the app from `feature/10058-check-if-product-already-blazed` with the new DB schema. I resolved that issue here: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2888/commits/b6df6f7a01d7971c416a0c074203f2671443d4c1 and with this other commit here: https://github.com/woocommerce/woocommerce-android/pull/10079/commits/048342d91cc72cdfd3442c549ff80215aa3a2be9

To test the fix, please: 
- Install he app from trunk and log into a site that has at least 1 Blaze campaign. 
- Then install the app from this branch and check it doesn't crash.

hichamboushaba commented:
> I've tried testing your PR but I'm seeing a weird empty state when initiating blaze. Do I need to enable something on the site? I'm getting the same behaviour on multiple sites.
> 
> I see this in the logs:
> 
> ```
> 2023-11-02 19:17:29.297  1032-1032  chromium                com.woocommerce.android.dev          I  [INFO:CONSOLE(28)] ?  "Uncaught TypeError: window.AppBoot is not a function", source: https://wordpress.com/advertising/woowoozela.com?blazepress-widget=post-7529&_source=my_store_section (28)
> 2023-11-02 19:17:29.297  1032-1032  chromium                com.woocommerce.android.dev          I  [INFO:CONSOLE(31)] "%cSTOP! color:#f00;font-size:xx-large", source: https://wordpress.com/advertising/woowoozela.com?blazepress-widget=post-7529&_source=my_store_section (31)
> 2023-11-02 19:17:29.297  1032-1032  chromium                com.woocommerce.android.dev          I  [INFO:CONSOLE(32)] "%cWait! This browser feature runs code that can alter your website or its security, and is intended for developers. If you've been told to copy and paste something here to enable a feature, someone may be trying to compromise your account. Please make sure you understand the code and trust the source before adding anything here. font-size:large;", source: https://wordpress.com/advertising/woowoozela.com?blazepress-widget=post-7529&_source=my_store_section (32)```

@0nko were you using the emulator for testing? I think Calypso has some issues with the emulator, and I face them also when trying to connect Jetpack sometime, but not on real devices, I think the Jetpack team faced some similar issues too p1694105449220099/1691408493.781129-slack-C04LJFS1G5P
Another possibility is that Calypso was facing some issues during the test.

JorgeMucientes commented:
Hey @hichamboushaba, just wanted to share that I faced the same issue as @onko shared in real devices. The way I can consistently reproduce this is: 
- Clean installation of the app, clearing all storage data. 
- Login with WP.com account
- Try to create a campaign, you'll experience the blank page. 

What is really interesting is, if you kill the app and retry creating a campaign again, everything will work. Check the screencast. I'm catching up with the thread p1694105449220099/1691408493.781129-slack-C04LJFS1G5P to see if it helps figure out what's happening. 

https://github.com/woocommerce/woocommerce-android/assets/2663464/5ea31b98-1726-424c-9b9d-60687fda7048

0nko commented:
@hichamboushaba Yes, I was using an emulator. When I used a real device, the first time I got a blank screen with "error" but then it worked the 2nd time, similar to what @JorgeMucientes mentioned.

hichamboushaba commented:
I can reproduce the same thing, and I was even able to reproduce it using the Jetpack app, still investigating to see if we can identify the cause, and see if it's still Calypso bug or a client issue.

0nko reviewed (approved):
I've tested the change and it works as expected, nice work! 👍 

Since the other issue I mentioned is not related, I'll go ahead and approve this PR.

JorgeMucientes committed:
Update FluxC changeset to reflect latest changes

0nko reviewed (approved):
LGTM 👍

JorgeMucientes committed:
Update FluxC changeset to latest trunk version
